lib/internal/modules/cjs/loader.js

```diff
@@ -1121,6 +1116,38 @@ Module._extensions['.js'] = function(module, filename) {
   } else {
     content = fs.readFileSync(filename, 'utf8');
   }
+  if (StringPrototypeEndsWith(filename, '.js')) {
+    const pkg = readPackageScope(filename);
+    // Function require shouldn't be used in ES modules.
+    if (pkg?.data?.type === 'module') {
+      const parent = moduleParentCache.get(module);
+      const parentPath = parent?.filename;
+      const packageJsonPath = path.resolve(pkg.path, 'package.json');
+      const usesEsm = hasEsmSyntax(content);
+      const err = new ERR_REQUIRE_ESM(filename, usesEsm, parentPath,
+                                      packageJsonPath);
+      // Attempt to reconstruct the parent require frame.
+      if (Module._cache[parentPath]) {
+        let parentSource;
+        try {
+          parentSource = fs.readFileSync(parentPath, 'utf8');
+        } catch {}
+        if (parentSource) {
+          const errLine = StringPrototypeSplit(StringPrototypeSlice(
+            err.stack, StringPrototypeIndexOf(err.stack, '    at ')), '\n')[0];
+          const { 1: line, 2: col } =
+              StringPrototypeMatch(errLine, /(\d+):(\d+)\)/) || [];
```

Comment:
`StringPrototypeMatch` relies on mutable built-ins (`RegExp.prototype[Symbol.match]`, `RegExp.prototype.exec`, `RegExp.prototype.global`, `RegExp.prototype.unicode`)
```suggestion
              RegExpPrototypeExec(/(\d+):(\d+)\)/, errLine) || [];
```